### PR TITLE
cosmic-greeter: init at unstable-2023-11-08

### PR DESCRIPTION
**Pull request by nyabinary**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by a-kenji** [comment]
Result of `nixpkgs-review pr 266138` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cosmic-greeter</li>
  </ul>
</details>

File: pkgs/by-name/co/cosmic-greeter/package.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      "cosmic-config-0.1.0" = "sha256-c2pGujYQ3WbbiHGhPo2kG8/NiydmpfFNQrlrb1nk/RY=";
+      "smithay-client-toolkit-0.17.0" = "sha256-vDY4cqz5CZD12twElUWVCsf4N6VO9O+Udl8Dc4arWK4=";
+      "softbuffer-0.2.0" = "sha256-VD2GmxC58z7Qfu/L+sfENE+T8L40mvUKKSfgLmCTmjY=";
+      "taffy-0.3.11" = "sha256-8gctP/nRiYxTSDrLyXi/oQbA7bE41ywgMbyotY1N8Zk=";
+    };
+  };
+
+  postPatch = ''
+    substituteInPlace justfile --replace '#!/usr/bin/env' "#!$(command -v env)"
+  '';
```
Comment (lilyinstarlight) [inline pkgs/by-name/co/cosmic-greeter/package.nix:45]: Is there a reason this is here? I don't see this substring (or anything like it) in `justfile` and when building I get a warning: `substituteStream(): WARNING: pattern '#!/usr/bin/env' doesn't match anything in file 'justfile'`

File: pkgs/by-name/co/cosmic-greeter/package.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  nativeBuildInputs = [ rustPlatform.bindgenHook cmake just pkg-config which lld util-linuxMinimal ];
```
Comment (lilyinstarlight) [inline pkgs/by-name/co/cosmic-greeter/package.nix:47]:
```suggestion
  nativeBuildInputs = [ rustPlatform.bindgenHook cmake just pkg-config ];
```

I can't tell that `util-linux`, `lld`, or `which` are used in the upstream build (and the debian `control` file doesn't list them as build dependencies), so they should probably be removed here

File: pkgs/by-name/co/cosmic-greeter/package.nix [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  buildInputs = [ libxkbcommon fontconfig freetype expat wayland linux-pam ];
```
Comment (lilyinstarlight) [inline pkgs/by-name/co/cosmic-greeter/package.nix:48]:
```suggestion
  buildInputs = [ libxkbcommon wayland linux-pam ];
```

It seems `fontconfig`, `freetype,` and `expat` aren't actually used here too

File: pkgs/by-name/co/cosmic-greeter/package.nix [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+, pkg-config
+, libxkbcommon
+, fontconfig
+, freetype
+, expat
+, linux-pam
+, wayland
+, which
+, lld
+, util-linuxMinimal
```
Comment (lilyinstarlight) [inline pkgs/by-name/co/cosmic-greeter/package.nix:17]:
```suggestion
, linux-pam
, wayland
```

Fixup args for above review comments

**Review by lilyinstarlight — COMMENTED** [review]
This looks good! I left a few minor nits for seemingly unused stuff, but otherwise this looks merge-ready

**Review by lilyinstarlight — APPROVED** [review]
The diff looks good to me now! Once the ofborg eval+builds all succeed, then I'll go ahead and merge, thanks!

**Comment by genofire** [comment]
How to use it?

the following does not work:
```nix
{pkgs, ...}:
{
  services.greetd = {
    enable = true;
    pkgs.greetd.cosmic-greeter;
  };
}
``

**Pull request merged**